### PR TITLE
add plugin gw-error-format

### LINKED ISSUE #62: Implement a plugin that can replace the response body according to the configuration
## Why do you need it?

实现一个插件，可以匹配应答和状态码返回自定义应答
---
Implement a plugin that can replace the response body according to the configuration

## How could it be?

config example:

```yaml
rules:
- match:
     status: 403
     body: "RBAC: access denied"
  replace:
     status: 200
     body: "{\"code\":403,\"message\":\"User is not authenticated\"}"
     content-type:  "application/json;utf-8"
- match:
     status: 503
     body: "no healthy upstream"
  replace:
     status: 200
     body: "{\"code\":503,\"message\":\"No Healthy Upstream\"}"
     content-type:  "application/json;utf-8"
```

### PR DESCRIPTION
**Pull request by zhangjiahaol**
### Ⅰ. Describe what this PR did
Implement a plugin that can replace the response body according to the configuration

### Ⅱ. Does this pull request fix one issue?

fixes #62

### Ⅲ. Why don't you add test cases (unit test/integration test)? 
No unit tests yet

### Ⅳ. Describe how to verify it
If you deployed higress as a K8S cluster, you can use this yaml to test replace-responsebody.yaml

### Ⅴ. Special notes for reviews
Check out README.md

### CODE REVIEW

File: plugins/wasm-go/extensions/gw-error-format/main.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+						if err != nil {
+							proxywasm.LogCritical("failed ReplaceHttpResponseHeader" + item_header.String())
+						}
+						return true
+					})
+				}
+			}
+			return types.ActionContinue
+		default:
+			DontReadResponseBody = true
```
Comment (johnlanni) [inline plugins/wasm-go/extensions/gw-error-format/main.go:83]: Is it also unnecessary to read the response body when the `x-envoy-upstream-service-time` header is present?

**Review by johnlanni — CHANGES_REQUESTED** [review]

File: plugins/wasm-go/extensions/gw-error-format/main.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+			} else {
+				ctx.DontReadResponseBody()
+			}
+			return types.ActionContinue
+		default:
+			DontReadResponseBody = true
+		}
+	}
+
+	if DontReadResponseBody == true {
```
Comment (johnlanni) [inline plugins/wasm-go/extensions/gw-error-format/main.go:88]: This bool here will never be false.

**Review by johnlanni — CHANGES_REQUESTED** [review]

File: plugins/wasm-go/extensions/gw-error-format/VERSION [hunk 3]
```diff
@@ -0,0 +1 @@
+1.0.0
```
Comment (Xunzhuo) [inline plugins/wasm-go/extensions/gw-error-format/VERSION:1]: nit: missing space at end

**Comment by johnlanni** [comment]
@zhangjiahaol Hello, do you have time to do some more optimizations on the code?

**Comment by zhangjiahaol** [comment]
> @zhangjiahaol Hello, do you have time to do some more optimizations on the code?

under fix.

File: plugins/wasm-go/extensions/gw-error-format/main.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+		switch current_statuscode {
+		//case "403", "503":
+		case config_match_statuscode:
+			// If the response header `x-envoy-upstream-service-time`  is not found,  the request has  not  been  forwarded to the  backend  service
+			_, err := proxywasm.GetHttpResponseHeader("x-envoy-upstream-service-time")
+			if err != nil {
+				proxywasm.AddHttpResponseHeader("config-match-statuscode", config_match_statuscode)
+				proxywasm.AddHttpResponseHeader("config-replace-statuscode", config_replace_statuscode)
+
+				proxywasm.RemoveHttpResponseHeader("content-length")
```
Comment (CH3CHO) [inline plugins/wasm-go/extensions/gw-error-format/main.go:63]: Do we need to handle errors returned by response header manipulating functions above?

File: plugins/wasm-go/extensions/gw-error-format/main.go [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+				// replace statuscode
+				err = proxywasm.ReplaceHttpResponseHeader(":status", config_replace_statuscode)
+				if err != nil {
+					log.Critical("failed ReplaceHttpResponseHeader :status")
```
Comment (CH3CHO) [inline plugins/wasm-go/extensions/gw-error-format/main.go:67]: Should we log the error itself?

**Review by CH3CHO — CHANGES_REQUESTED** [review]

File: plugins/wasm-go/extensions/gw-error-format/main.go [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+	}
+
+	return nil
+}
+
+func onHttpResponseHeader(ctx wrapper.HttpContext, config MyConfig, log wrapper.Log) types.Action {
+	var DontReadResponseBody = false
+	// judge statuscode
+	for _, item := range config.rules {
+		current_statuscode, err := proxywasm.GetHttpResponseHeader(":status")
```
[the remark this answers is not in the document]
Comment (zhangjiahaol) [reply plugins/wasm-go/extensions/gw-error-format/main.go:47]: fix

**Review by CH3CHO — APPROVED** [review]
LGTM

**Pull request merged**